add var skip_assert_errors & var etcd-even-numbers to allow an even number of etcd nodes

pedro-peter opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md and developer guide https://git.k8s.io/community/contributors/devel/development.md
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/guide/release-notes.md
6. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What type of PR is this?**
/kind feature
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespaces from that line:
>
> /kind api-change
> /kind bug
> /kind cleanup
> /kind design
> /kind documentation
> /kind failing-test
> /kind flake

**What this PR does / why we need it**:
Currently the only way to add an even number of control-plane/etcd nodes to a cluster is to disable all assert checks by passing `ignore_assert_errors=yes`

It would be better just to skip the etcd even nodes assert check and leave all other asserts as they can catch useful problems.

**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #12013

**Special notes for your reviewer**:
I've updated the docs to use the new `-e '{"skip_assert_errors": ["etcd-even-numbers"]}'` flag.  This isn't really a breaking change as  `ignore_assert_errors=yes` will still work.

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".
-->
```release-note
add var skip_assert_errors & var etcd-even-numbers to allow an even number of etcd nodes
```

ant31 commented:
In which situation would you want even number of etcd?

pedro-peter commented:
> In which situation would you want even number of etcd?

when replacing a control-plane/etcd node, you might want to add the new node before removing the old node:
https://github.com/kubernetes-sigs/kubespray/blob/master/docs/operations/nodes.md#adding-an-etcd-node

there's a short duration where you'll have an even number of etcd nodes.

ant31 commented:
yes, it's remove then add
 
Etcd documentation on live reconfigurations:
https://etcd.io/docs/v3.5/op-guide/runtime-configuration/

> To replace the machine, follow the instructions for [removing the member](https://etcd.io/docs/v3.5/op-guide/runtime-configuration/#remove-a-member) from the cluster, and then [add a new member](https://etcd.io/docs/v3.5/op-guide/runtime-configuration/#add-a-new-member) in its place. If the cluster holds more than 50MB, it is recommended to [migrate the failed member’s data directory](https://etcd.io/docs/v2.3/admin_guide/#member-migration) if it is still accessible

and 
>All changes to the cluster must be done sequentially:
To update a single member peerURLs, issue an update operation
To replace a healthy single member, remove the old member then add a new member
To increase from 3 to 5 members, issue two add operations
To decrease from 5 to 3, issue two remove operations

Also I don't think kubespray etcd role is tested to do such operations.

Overall I don't think we should add variable to allow an even number of etcd. It's never a situation you want to be in and could mislead users new to kubernetes.

pedro-peter commented:
I still think it's useful to allow creating an uneven number of nodes (as long as it's disabled by default) 

currently you can only do this by disabing all assert checks...

cluster operators can make the decsion if they want to risk running a cluster for a duration with an uneven number of etcd nodes.

morguldir commented:
When you remove a node you also end up with an even amount of nodes, and with 2 nodes the quorum is 2, so until you deploy the new node, you have zero failure tolerance

But I think the safer way to avoid this is to add two nodes, e.g. etcd4 and etcd5, then remove etcd3 and etcd5 at the same time. Using etcd learners would also help

pedro-peter commented:
I think people are missing the point of this PR...

you can already add an uneven number of control-plane/etcd nodes in a kubespray cluster.  You currently need to disable *all*  assert checks (https://github.com/kubernetes-sigs/kubespray/blob/master/roles/kubernetes/preinstall/tasks/0040-verify-settings.yml)  using var `ignore_assert_errors=yes` to do this (this is in the documentation).

If folks want to add an even number of control-plane/etcd nodes it's better just to allow disabling the single asert check (https://github.com/kubernetes-sigs/kubespray/pull/12014/files#diff-f817513f21cf14346b8672affc950686b1c4e3b9521671e0bbb76aac63fa5f6bR67) and allow the remaining checks to run.

morguldir commented:
I don't think anyone missed that, adding an even amount of etcd nodes is a large footgun, so i understand not wanting to add any form of support for it

> this is in the documentation

It probably shouldn't be, but it's at least obvious that it's a dangerous operation

pedro-peter commented:
> But I think the safer way to avoid this is to add two nodes, e.g. etcd4 and etcd5, then remove etcd3 and etcd5 at the same time. Using etcd learners would also help

this is definitely a safer operation to replace a control-plane node should probably be added to the documentation:
https://github.com/kubernetes-sigs/kubespray/blob/master/docs/operations/nodes.md#1-add-the-new-node-running-clusteryml

there are cases though where a cluster operator might decide to take the risk of running an even number of etcd node (for a short duration).  folks may not have the hardware available to spin up new control-plane nodes.

VannTen commented on inventory/sample/group_vars/all/all.yml:
```diff
@@ -88,6 +88,10 @@ no_proxy_exclude_workers: false
 ## Set to true to allow pre-checks to fail and continue deployment
 # ignore_assert_errors: false
 
+## Set to true to skip 'Stop if even number of etcd hosts' assert check
+## Running an even number of etcd nodes is *not* recommended for production clusters
+skip_etcd_even_nodes_check: false
+
```
Remove this. The inventory sample is for "typical" settings, and not only this should not be a common case, this should never be in the inventory, probably only specified in extra_var.

VannTen commented on roles/kubernetes/preinstall/defaults/main.yml:
```diff
@@ -1,8 +1,12 @@
 ignore_assert_errors: false
+# Set to true to skip 'Stop if even number of etcd hosts' assert check
+# Running an even number of etcd nodes is *not* recommended for production clusters
+skip_etcd_even_nodes_check: false
```
Point to consider: since it shouldn't be in inventory, maybe we should put it in vars instead of default so inventory does not override it

VannTen commented on roles/kubernetes/preinstall/defaults/main.yml:
```diff
@@ -1,8 +1,12 @@
+skip_etcd_even_nodes_check: false
 # Set to false to disable the backup parameter, set to true to accumulate backups of config files.
 leave_etc_backup_files: true
+
```
Drop this (avoid whitespaces only change in general especially if unrelated)

VannTen commented on roles/kubespray-defaults/defaults/main/main.yml:
```diff
@@ -80,6 +80,10 @@ kube_proxy_nodeport_addresses: >-
 # Set to true to allow pre-checks to fail and continue deployment
 ignore_assert_errors: false
 
+# Set to true to skip 'Stop if even number of etcd hosts' assert check
+# Running an even number of etcd nodes is *not* recommended for production clusters
+skip_etcd_even_nodes_check: false
+
```
The variable should only be in one place.

VannTen reviewed (commented):
I don't have a strong opinion on this, but some style comments below:

pedro-peter commented:
Thanks @VannTen, I've rebased and updated the commit with your suggestions.

The `Stop if even number of etcd hosts` assert check got moved to a new role `validate_inventory` in https://github.com/kubernetes-sigs/kubespray/commit/1b9919547ab9465dc7cfe11d81b2a36edc8838f1#diff-0a24eb7611eb3fe53e9bfd81fd8ffc1545c23a1827c7e93efcfbaa1af943654bR97

ant31 commented:
If we go this path, we should do a more generic way to skip checks (like for kubeadm):
```
skip_assert_errors: 
  - etcd-even-numbers
  - XYZ
  - BLA
``` 
```
when:
    - not 'etcd-even-numbers' in skip_assert_errors |d([])
```

that way we can add more skips in a granular way without 100s more variable to configure

yankay commented:
/ok-to-test
